python/ray/data/datasource/tfrecords_datasource.py

```diff
@@ -175,27 +175,79 @@ def _value_to_feature(value: Union[bytes, float, int, List]) -> "tf.train.Featur
-        if file.readinto(crc_bytes) != 4:
-            raise ValueError("Failed to read the start token.")
-        (length,) = struct.unpack("<Q", length_bytes)
-        if length > len(datum_bytes):
-            datum_bytes = datum_bytes.zfill(int(length * 1.5))
-        datum_bytes_view = memoryview(datum_bytes)[:length]
-        if file.readinto(datum_bytes_view) != length:
-            raise ValueError("Failed to read the record.")
-        if file.readinto(crc_bytes) != 4:
-            raise ValueError("Failed to read the end token.")
-        yield datum_bytes_view
+        try:
+            # Read "length" field.
+            num_length_bytes_read = file.readinto(length_bytes)
+            if num_length_bytes_read == 0:
+                break
+            elif num_length_bytes_read != 8:
+                raise ValueError(
+                    "Failed to read the length of record data. Expected 8 bytes but "
+                    "got {num_length_bytes_read} bytes."
+                )
+
+            # Read "masked_crc32_of_length" field.
+            num_length_crc_bytes_read = file.readinto(crc_bytes)
+            if num_length_crc_bytes_read != 4:
+                raise ValueError(
+                    "Failed to read the length of CRC-32C hashes. Expected 4 bytes "
+                    "but got {num_length_crc_bytes_read} bytes."
+                )
+
+            # Read "data[length]" field.
+            (data_length,) = struct.unpack("<Q", length_bytes)
+            if data_length > len(datum_bytes):
+                datum_bytes = datum_bytes.zfill(int(data_length * 1.5))
+            datum_bytes_view = memoryview(datum_bytes)[:data_length]
+            num_datum_bytes_read = file.readinto(datum_bytes_view)
+            if num_datum_bytes_read != data_length:
+                raise ValueError(
+                    f"Failed to read the record. Exepcted {data_length} bytes but got "
+                    f"{num_datum_bytes_read} bytes."
+                )
+
+            # Read "masked_crc32_of_data" field.
+            # TODO(chengsu): ideally we should check CRC-32C against the actual data.
+            num_crc_bytes_read = file.readinto(crc_bytes)
+            if num_crc_bytes_read != 4:
+                raise ValueError(
+                    "Failed to read the CRC-32C hashes. Expected 4 bytes but got "
+                    f"{num_crc_bytes_read} bytes."
+                )
+
+            # Return the data.
+            yield datum_bytes_view
+
+            row_count += 1
+            data_length = None
+        except Exception:
+            error_message = (
+                f"Failed to read TFRecord file {path}. Please check the TFRecord file "
+                f"has correct format. Already read {row_count} rows."
+            )
+            if data_length is not None:
+                error_message += f" Bytes size of current record data is {data_length}."
+            raise Exception(error_message)
```

Comment:
A few nits for this exception handling:
1. Small exception message nits.
2. We should (1) raise a `RuntimeError` here instead of a generic exception.
3. We should use `from e` exception chaining to make the relationship of these two exceptions clear.
```
{{caught error traceback}}
The above exception was the direct cause of the following exception:
{{outer failed TFRecord read error traceback}}
```
Just doing `raise RuntimeError(error_message)` without the chaining would result in
```
{{caught error traceback}}
During handling of the above exception, another exception occurred:
{{outer failed TFRecord read error traceback}}
```
which makes it look like we were trying to handle `e` but failed, with a different error being raised.
```suggestion
        except Exception as e:
            error_message = (
                f"Failed to read TFRecord file {path}. Please ensure that the TFRecord file "
                f"has correct format. Already read {row_count} rows."
            )
            if data_length is not None:
                error_message += f" Byte size of current record data is {data_length}."
            raise RuntimeError(error_message) from e
```

Reply:
Thanks @clarkzinzow, updated.